### PR TITLE
added simple sss-create and sss-restore command-line tools

### PR DESCRIPTION
**Pull request by stef**
they can be invoked by strictly following the simple example:
```
echo -n "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef" | ./sss-create 5 4 test && cat test*[4230]* | ./sss-restore 4
```

TODO: if upstream accepts this: documentation.
      also cleaning of the memory of sensitive data*, if someone intends this to use for real
	   (sodium_memzero means dep to libsodium)

### CODE REVIEW

**Comment by stef** [comment]
hmmm there seems to be something wrong with your travis setup i guess?

**Comment by dsprenkels** [comment]
Hey @stef, there is already a command line interface implemented here: https://github.com/dsprenkels/sss-cli

It is in a separate repository and written in Rust (because of memory safety). I try to keep the actual crypto code base as small as possible, (1) to make it easier for people to audit and (2) to keep the attack surface as small as possible.

I'd be happy to accept issues and PR's in [sss-cli](https://github.com/dsprenkels/sss-cli) for improvements.

About `memzero`'ing, that's not quite trivial. If we *really* want to memzero everything properly, we will have to explicitly set not only the buffers to `'\0'`, but also every variable on the stack and register, after it has been used. I really rather just trust the OS to isolate the memory pages properly.

**Comment by stef** [comment]
> Hey @stef, there is already a command line interface implemented here: https://github.com/dsprenkels/sss-cli

> It is in a separate repository and written in Rust (because of memory safety). 

yeah, that's why i wrote this in C, because rust is not very portable (i'm on a distro that still has problems packaging rust). Also i do not see how memory safety is a problem in this particular case. what exactly is your threat model where you think memory safety is an important issue?

> I try to keep the actual crypto code base as small as possible, (1) to make it easier for people to audit and (2) to keep the attack surface as small as possible.
 
makes sense.

> I'd be happy to accept issues and PR's in sss-cli for improvements.

i thought it is nice to have some simple c cli programs without any dependencies with which to test the lib. Do you think it is a good idea to have a c and rust cli in the same pakage? if so i can send a PR to sss-cli, otherwise i just make my own repo.

> About memzero'ing, that's not quite trivial. If we really want to memzero everything properly, we will have to explicitly set not only the buffers to '\0', but also every variable on the stack and register, after it has been used. I really rather just trust the OS to isolate the memory pages properly.

That is a nice assumption, but on linux the kernel only zeroes memory if allocated to a new process, if there is (and there are lots of) infoleaks then the sensitive data can be still accessed after termination of processes. only the pax sanitize patch does immediate sanitization, but pax/grsecurity went proprietary. Besides the newer linuxen there's also other OSes that are not quite to be trusted with sanitizing memory correctly, especially if we go back to older kernels used on some embedded systems or appliances. Note i also intended to clean the stack, but i do not see in my code where the sanitization of registers makes any difference, i'll investigate that.

**Comment by dsprenkels** [comment]
> yeah, that's why i wrote this in C, because rust is not very portable (i'm on a distro that still has problems packaging rust).

I am wondering. What platform are you on? I am interested in porting `sss-cli` to as many platforms as possible. :)

> Also i do not see how memory safety is a problem in this particular case. what exactly is your threat model where you think memory safety is an important issue?

Well. It's exactly the "attack surface" argument. Programmes make mistakes. In C these mistakes can lead to buffer overflows and all sorts of icky memory bugs. If low level memory access or language interoperability is not required, then I really like to implement my security sensitive software in a language that enforces type-safety.

> i thought it is nice to have some simple c cli programs without any dependencies with which to test the lib. Do you think it is a good idea to have a c and rust cli in the same pakage? if so i can send a PR to sss-cli, otherwise i just make my own repo.

I am not sure what your plans are. Both is probably fine, although t may be easier to just put it into your own repo. In any case, I'd be happy to help with stuff (and put links in proper `README.md` files).

> That is a nice assumption, [...]. Note i also intended to clean the stack, but i do not see in my code where the sanitization of registers makes any difference, i'll investigate that.

There is no problem with your code. :)

It's just that this repository assumes that the OS is secure and enforces proper memory isolation. If it doesn't, then it's a bug in the OS. The reason behind this reasoning is the fact that it is practically impossible to get proper randomness in userspace. We have to get our randomness from the kernel (see the `randombytes` submodule). The OS better be secure.

In other words: So many things can go wrong if the OS is insecure. The randomness could be bad; there may be side channels in the kernel leaking userspace memory; free'd memory or registers could be leaked; some user could unintendedly have gained arbitrary memory access over the process. From userspace, there is no way that we can protect against this all. So we just assume the OS doesn't suck, and as such we don't have to clear our memories.

**Comment by stef** [comment]
> I am wondering. What platform are you on? I am interested in porting sss-cli to as many platforms as possible. :)

alpine linux, if you want to make this portable, you have to bootstrap rust with a c compiler.

> > Also i do not see how memory safety is a problem in this particular case. what exactly is your threat model where you think memory safety is an important issue?

> Well. It's exactly the "attack surface" argument. Programmers make mistakes. In C these mistakes can lead to buffer overflows and all sorts of icky memory bugs. If low level memory access or language interoperability is not required, then I really like to implement my security sensitive software in a language that enforces type-safety.

that is a very generic answer without considering the concrete implementation (no buffer overflows possible, no icky memory bugs (except infoleaks of sensitive data) and the threatmodell where the user itself is running a one shot application that cannot lead to neither code execution nor privilege escalation.

> I am not sure what your plans are. Both is probably fine, although t may be easier to just put it into your own repo. In any case, I'd be happy to help with stuff (and put links in proper README.md files). :D

ack, i'll do my own repo. no problems.

> > That is a nice assumption, [...]. Note i also intended to clean the stack, but i do not see in my code where the sanitization of registers makes any difference, i'll investigate that.

> There is no problem with your code. :)

thanks. ;)

> It's just that this repository assumes that the OS is secure and enforces proper memory isolation. If it doesn't, then it's a bug in the OS. The reason behind this reasoning is the fact that it is practically impossible to get proper randomness in userspace. We have to get our randomness from the kernel (see the randombytes submodule). The OS better be secure.

that is a nice assumption but unfortunately a fallacy, most OSes are still incredibly bad at this kind of keeping userland sensitive data secure.

> In other words: So many things can go wrong if the OS is insecure. The randomness could be bad; there may be side channels in the kernel leaking userspace memory; free'd memory or registers could be leaked; some user could unintendedly have gained arbitrary memory access over the process. From userspace, there is no way that we can protect against this all. So we just assume the OS doesn't suck, and as such we don't have to clear our memories.

clearing sensitive data where possible would be a nice defense-in-depth though, since you already care about other attacks that the other ssss implementations ignore.
